### ISSUE TITLE
latest package:ffi causes test failures on the SDK bots

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/ffi#148]
We're seeing some test failures when trying to roll the latest package:ffi to the SDK repo. Some details:

- the package had been up-to-date modulo the commit from the last PR: https://github.com/dart-lang/ffi/pull/144
- the test failures on are windows

Some more info re: the failures are in these links: https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8811167916042190129/+/u/test_results/stdout and https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8811167916042190129/+/u/test_results/new_test_failures__logs_. I'll in-line here in case those links expire:

```
/===============================================================\
| ffi_2/data_not_asan_test broke (Pass -> Crash, expected Pass) |
\===============================================================/

--- Command "vm_compile_to_kernel" (took 04.000033s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & C:\b\s\w\ir\pkg\vm\tool\gen_kernel.bat --aot --platform=out/ReleaseX64/vm_platform_strong.dill -o C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\tests_ffi_2_data_not_asan_test\out.dill C:\b\s\w\ir\tests\ffi_2\data_not_asan_test.dart -Dtest_runner.configuration=dartkp-win-release-x64 --packages=C:\b\s\w\ir\.dart_tool\package_config.json -Ddart.vm.product=false

exit code:
0

--- Command "precompiler" (took 04.000934s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & out\ReleaseX64\gen_snapshot --snapshot-kind=app-aot-elf --elf=C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/tests_ffi_2_data_not_asan_test/out.aotsnapshot --loading-unit-manifest=C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/tests_ffi_2_data_not_asan_test/ignored.json -Dtest_runner.configuration=dartkp-win-release-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.dart_tool\package_config.json C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\tests_ffi_2_data_not_asan_test\out.dill

exit code:
0

--- Command "remove_kernel_file" (took 252ms):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & cmd.exe /c del C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\tests_ffi_2_data_not_asan_test\out.dill

exit code:
0

--- Command "vm" (took 476ms):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & out\ReleaseX64\dart_precompiled_runtime.exe -Dtest_runner.configuration=dartkp-win-release-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.dart_tool\package_config.json C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/tests_ffi_2_data_not_asan_test/out.aotsnapshot

exit code:
-1073741819

--- Re-run this test:
python3 tools/test.py -n dartkp-win-release-x64 ffi_2/data_not_asan_test


/=================================================================\
| samples_2/ffi/samples_test broke (Pass -> Crash, expected Pass) |
\=================================================================/

--- Command "vm_compile_to_kernel" (took 03.000862s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & C:\b\s\w\ir\pkg\vm\tool\gen_kernel.bat --aot --platform=out/ReleaseX64/vm_platform_strong.dill -o C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\samples_2_ffi_samples_test\out.dill C:\b\s\w\ir\samples_2\ffi\samples_test.dart -Dtest_runner.configuration=dartkp-win-release-x64 --packages=C:\b\s\w\ir\.dart_tool\package_config.json -Ddart.vm.product=false

exit code:
0

--- Command "precompiler" (took 05.000134s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & out\ReleaseX64\gen_snapshot --snapshot-kind=app-aot-elf --elf=C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/samples_2_ffi_samples_test/out.aotsnapshot --loading-unit-manifest=C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/samples_2_ffi_samples_test/ignored.json -Dtest_runner.configuration=dartkp-win-release-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.dart_tool\package_config.json C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\samples_2_ffi_samples_test\out.dill

exit code:
0

--- Command "remove_kernel_file" (took 51ms):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & cmd.exe /c del C:\b\s\w\ir\out\ReleaseX64\generated_compilations\dartkp-win-release-x64\samples_2_ffi_samples_test\out.dill

exit code:
0

--- Command "vm" (took 697ms):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\ReleaseX64\crashpad_handler.exe & out\ReleaseX64\dart_precompiled_runtime.exe -Dtest_runner.configuration=dartkp-win-release-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.dart_tool\package_config.json C:/b/s/w/ir/out/ReleaseX64/generated_compilations/dartkp-win-release-x64/samples_2_ffi_samples_test/out.aotsnapshot

exit code:
-1073741819

stdout:
0
1
15
1
0
0
2
start main
Pointer<Int64> value: 42
If not set, returns garbage: 0
Address: 2134826018400
Pointer<Int64> value: 1337
After free, address is zero: 2134826018192
Expected exception on allocating too much

--- Re-run this test:
python3 tools/test.py -n dartkp-win-release-x64 samples_2/ffi/samples_test

```

cc @dcharkes @timsneath

**Comment by dcharkes** [comment]
@timsneath that must be the changes to the allocator you made, could you please take a look?

**Comment by timsneath** [comment]
For the samples_test, I think this _might_ be OK (or at least, by design according to the Windows documentation). It looks like the test is asserting that only the exact number of bytes asked for is allocated. But from [the `CoTaskMemAlloc` docs](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-cotaskmemalloc):

> The allocated block may be larger than `cb` bytes because of the space required for alignment and for maintenance information.

For the data_not_asan_test, this looks like a bug in that we try to zero out the memory even if the calloc call fails because not enough memory could be allocated. Submitting a PR for that one to the package:ffi repo.

**Comment by dcharkes** [comment]
Closed by #150
